docs: #842: create new Schema Registry topic with blog post content and API reference link (#899)
redpanda-data/documentation: docs/platform/data-management/schema-registry.mdx

---
title: Schema Registry
---

<head>
    <meta name="title" content="Redpanda Schema Registry | Redpanda Docs"/>
    <meta name="description" content="Overview, concepts, and tutorials of the schema registry in Redpanda."/>
</head>

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

Redpanda's schema registry provides the interface to store and manage event schemas:

- Query supported serialization formats.
- Register schemas for a subject.
- Retrieve schemas of specific versions.
- Configure schema compatibilty checking.

> Redpanda built the schema registry directly into the Redpanda binary. The schema registry runs out-of-the-box with Redpanda's default configuration, and it requires no new binaries to install and no new services to deploy or maintain.

Author: Kafka users usually expect additional configuration for a schema registry. Making this statement in the overview preemptively answers this frequently asked question.

> Redpanda built the schema registry directly into the Redpanda binary. The schema registry runs out-of-the-box with Redpanda's default configuration, and it requires no new binaries to install and no new services to deploy or maintain.

Author: After more thought, this is a detail specific to Redpanda's implementation that fits better in a `Redpanda schema registry` or `Redpanda implementation` explanation, which I'll add.

> The schema registry supports different serialization formats:

Author: To have a short, scannable overview, how about editing this to `schema serialization format`, and adding a link to a more detailed concept/explanation further down?

> The schema registry supports different serialization formats:

Author: After more thought, this fits better under a `Serialization` sub-heading, and because the overview and concepts sections are short, it will still be scannable after moving it under concepts.

- Avro

> - Protocol buffers 

Reviewer: Should we add protobuf in parentheses?

## Concepts

A schema is an external mechanism to describe the structure of data and its encoding.  

Using schemas provides multiple benefits:
- Enables a loosely coupled, data centric architecture that minimizes dependencies in code, between teams, and between producers and consumers.
- Provides human-readable documentation for an API.
- Verifies that data conforms to an API.
- Supports the generation of serializers for data.
- Manages the compatibility of evolving APIs, allowing new versions of services to be rolled out independently.

> By defining an API, or in this case the event, with a schema, disparate teams can start work on the subsystems that produce and consume those events with minimal communication overhead.

Author: Agreed, will remove.

> Schemas are stored in a standard compacted topic. Redpanda uses optimistic concurrency control at the topic level to allow mutating REST calls to any broker. Every broker is symmetric, so there is no need to configure leadership or failover strategies.

Author: Schema registry updates are done through mutating REST calls. Updating the sentence to provide this context by appending `to update the schema registry.`

> ## Tutorial

Author: Raising this to the team, I like `Tutorial` for a section with both explanations and steps, but for concistency we all should agree on the terminology.

> ## Tutorial

Author: Changing to `Examples`

This tutorial covers the basic functionality of the Redpanda schema registry with an example Avro schema, `sensor_sample`, that represents a measurement from a sensor for the value of the `sensor` topic:

```json
{
  "type": "record",
  "name": "sensor_sample",
  "fields": [
    {
      "name": "timestamp",
      "type": "long",
      "logicalType": "timestamp-millis"
    },
    {
      "name": "identifier",
      "type": "string",
      "logicalType": "uuid"
    },
    {
      "name": "value",
      "type": "long"
    }
  ]
}
```

### Prerequisites

1. Download the [jq utility](https://stedolan.github.io/jq/download/). 
2. Install [curl](https://curl.se/) or [Python](https://www.python.org/).
    - If using Python, install the [Requests module](https://requests.readthedocs.io/en/latest/user/install/#install), then create an interactive Python session:

        ```python
        import requests
        import json
        def pretty(text):
          print(json.dumps(text, indent=2))

        base_uri = "http://localhost:8081"
        ```

> 3. [Install Redpanda](/docs/platform/quickstart/index.mdx). This tutorial uses [Redpanda with Docker](/docs/platform/quickstart/quick-start-docker).

Author: Agreed

4. Start Redpanda with Docker:

    ```bash
    docker network create redpanda-sr
    docker volume create redpanda-sr
    docker run \
    --pull=always \
    --name=redpanda-sr \
    --net=redpanda-sr \
    -v "redpanda-sr:/var/lib/redpanda/data" \
    -p 8081:8081 \
    -p 8082:8082 \
    -p 9092:9092 \
    --detach \
    docker.vectorized.io/vectorized/redpanda start \
    --overprovisioned \
    --smp 1 \
    --memory 1G \
    --reserve-memory 0M \
    --node-id 0 \
    --check=false \
    --pandaproxy-addr 0.0.0.0:8082 \
    --advertise-pandaproxy-addr 127.0.0.1:8082 \
    --kafka-addr 0.0.0.0:9092 \
    --advertise-kafka-addr redpanda-sr:9092
    ```

    - Once Redpanda is running, endpoints are documented at `http://localhost:8081/v1`.

### Query supported schema formats

To get the supported data serialization formats in the schema registry, make a GET request to the `/schemas/types` endpoint:

<Tabs>
<TabItem value="curl" label="Curl" default>

```bash
curl -s "http://localhost:8081/schemas/types" | jq .
```

</TabItem>
<TabItem value="python" label="Python">

```python
res = requests.get(f'{base_uri}/schemas/types').json()
pretty(res)
```

</TabItem>
</Tabs>

> On success, it returns the supported serialization formats:

Author: I agree on changing to `If the request is successful`. I don't agree on adding `For example:`, because the documented output is from the specific request the user made. 

(My comment applies to all similar comments further down this file.)

- `FORWARD_TRANSITIVE` - consumers using any previous schema (for example, versions 1-9) can read data from producers using the new schema (for example, version 10).
- `FULL` - the new schema is compatible with all previous schemas.
- `NONE` - no schema compatiblity checks are done.

To set the compatibility type for a subject, make a PUT request to `/config/<subject-name>` with the specific compatibility type:

<Tabs>
<TabItem value="curl" label="Curl" default>

```bash
curl -s \
  -X PUT \
  "http://localhost:8081/config/sensor-value" \
  -H "Content-Type: application/vnd.schemaregistry.v1+json" \
  -d '{"compatibility": "BACKWARD"}' \
  | jq .
```

</TabItem>
<TabItem value="python" label="Python">

```python
res = requests.put(
    url=f'{base_uri}/config/sensor-value',
    data=json.dumps(
        {'compatibility': 'BACKWARD'}
      ),
    headers={'Content-Type': 'application/vnd.schemaregistry.v1+json'}).json()
pretty(res)
```

</TabItem>
</Tabs>

On success, it returns the new compatibility type:

```json
{
  "compatibility": "BACKWARD"
}
```

> Posting an incompatible schema change returns an error. For example, given our `BACKWARD` compatible subject `sensor-value`, registering a new schema with its `value` field's type changed from `long` to `int` returns an error due to incompatibility:

Reviewer 2: ```suggestion
If you POST an incompatible schema change, the request returns an error. For example, if you try to register a new schema with the `value` field's type changed from `long` to `int`, and compatibility is set to `BACKWARD`,  the request returns an error due to incompatibility:
```

<Tabs>
<TabItem value="curl" label="Curl" default>

```bash
curl -s \
  -X POST \
  "http://localhost:8081/subjects/sensor-value/versions" \
  -H "Content-Type: application/vnd.schemaregistry.v1+json" \
  -d '{"schema": "{\"type\":\"record\",\"name\":\"sensor_sample\",\"fields\":[{\"name\":\"timestamp\",\"type\":\"long\",\"logicalType\":\"timestamp-millis\"},{\"name\":\"identifier\",\"type\":\"string\",\"logicalType\":\"uuid\"},{\"name\":\"value\",\"type\":\"int\"}]}"}' \
  | jq
```

</TabItem>
<TabItem value="python" label="Python">

```python
sensor_schema["fields"][2]["type"] = "int"

res = requests.post(
    url=f'{base_uri}/subjects/sensor-value/versions',
    data=json.dumps({
      'schema': json.dumps(sensor_schema)
    }),
    headers={'Content-Type': 'application/vnd.schemaregistry.v1+json'}).json()
pretty(res)
```

</TabItem>
</Tabs>

> The returned error:

Reviewer 2: ```suggestion
The request returns this error:
```

  "message": "Schema being registered is incompatible with an earlier schema for subject \"{sensor-value}\""
}
```

For an example of a compatible change, register a schema with the `value` field's type changed from `long` to `double`:

<Tabs>
<TabItem value="curl" label="Curl" default>

```bash
curl -s \
  -X POST \
  "http://localhost:8081/subjects/sensor-value/versions" \
  -H "Content-Type: application/vnd.schemaregistry.v1+json" \
  -d '{"schema": "{\"type\":\"record\",\"name\":\"sensor_sample\",\"fields\":[{\"name\":\"timestamp\",\"type\":\"long\",\"logicalType\":\"timestamp-millis\"},{\"name\":\"identifier\",\"type\":\"string\",\"logicalType\":\"uuid\"},{\"name\":\"value\",\"type\":\"double\"}]}"}' \
  | jq
```

</TabItem>
<TabItem value="python" label="Python">

```python
sensor_schema["fields"][2]["type"] = "double"

res = requests.post(
    url=f'{base_uri}/subjects/sensor-value/versions',
    data=json.dumps({
      'schema': json.dumps(sensor_schema)
    }),
    headers={'Content-Type': 'application/vnd.schemaregistry.v1+json'}).json()
pretty(res)
```

</TabItem>
</Tabs>

A successful registration returns the schema's `id`:

```json
{
  "id": 2
}
```

### Cleanup 

> When done with the tutorial, stop and cleanup Redpanda:

Reviewer 2: ```suggestion
When you're finished , stop and clean up Redpanda:
```

---
title: Schema Registry
---

<head>
    <meta name="title" content="Redpanda Schema Registry | Redpanda Docs"/>
    <meta name="description" content="Overview, concepts, and tutorials of the schema registry in Redpanda."/>
</head>

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

Redpanda's schema registry provides the interface to store and manage event schemas. Producers register the schemas of data they send, and consumers retrieve the schemas of data they receive. The registry manages the compatibility between different versions of schemas, and it notifies producers and consumers when they exchange data with incompatible schemas.

Its API endpoints support the following functionality: 

- Querying supported serialization formats.
- Registering schemas for a subject.

> - Retrieving schemas of specific versions.

Reviewer 3: Additionally:
```
- Retrieving a list of subjects.
- Retrieving a list of schema versions for a subject.
```

1. Download the [jq utility](https://stedolan.github.io/jq/download/). 
2. Install [curl](https://curl.se/) or [Python](https://www.python.org/).
    - If using Python, install the [Requests module](https://requests.readthedocs.io/en/latest/user/install/#install), then create an interactive Python session:

        ```python
        import requests
        import json
        def pretty(text):
          print(json.dumps(text, indent=2))

        base_uri = "http://localhost:8081"
        ```
3. [Install Redpanda with Docker](/docs/platform/quickstart/quick-start-docker).
4. Start Redpanda with Docker:

    ```bash
    docker network create redpanda-sr
    docker volume create redpanda-sr
    docker run \
    --pull=always \
    --name=redpanda-sr \
    --net=redpanda-sr \
    -v "redpanda-sr:/var/lib/redpanda/data" \
    -p 8081:8081 \
    -p 8082:8082 \
    -p 9092:9092 \
    --detach \
    docker.vectorized.io/vectorized/redpanda start \
    --overprovisioned \
    --smp 1 \
    --memory 1G \
    --reserve-memory 0M \
    --node-id 0 \
    --check=false \
    --pandaproxy-addr 0.0.0.0:8082 \
    --advertise-pandaproxy-addr 127.0.0.1:8082 \
    --kafka-addr 0.0.0.0:9092 \

>     --advertise-kafka-addr redpanda-sr:9092

Reviewer 3: Some of these flags are probably not needed any more since we now have `rpk redpanda start --mode dev-container`, but i haven't played around with that yet.

Author: Created #929 to address this.

```bash
curl -s \
  -X POST \
  "http://localhost:8081/subjects/sensor-value/versions" \
  -H "Content-Type: application/vnd.schemaregistry.v1+json" \
  -d '{"schema": "{\"type\":\"record\",\"name\":\"sensor_sample\",\"fields\":[{\"name\":\"timestamp\",\"type\":\"long\",\"logicalType\":\"timestamp-millis\"},{\"name\":\"identifier\",\"type\":\"string\",\"logicalType\":\"uuid\"},{\"name\":\"value\",\"type\":\"double\"}]}"}' \
  | jq
```

</TabItem>
<TabItem value="python" label="Python">

```python
sensor_schema["fields"][2]["type"] = "double"

res = requests.post(
    url=f'{base_uri}/subjects/sensor-value/versions',
    data=json.dumps({
      'schema': json.dumps(sensor_schema)
    }),
    headers={'Content-Type': 'application/vnd.schemaregistry.v1+json'}).json()
pretty(res)
```

</TabItem>
</Tabs>

A successful registration returns the schema's `id`:

```json
{
  "id": 2
}
```

### Cleanup 

When you're finished, stop and clean up Redpanda:

```bash
docker stop redpanda-sr
docker rm redpanda-sr
docker volume remove redpanda-sr
docker network remove redpanda-sr
```

## Related topics

- [Schema Registry API reference](../../reference/schema-registry-api)

> - [Sample Redpanda node configuration file](../../reference/node-configuration-sample) (search for `schema_registry`)

Reviewer 3: It would be really nice if all configuration properties had an anchor. I generally end up sending people links like this: [schema_registry](https://deploy-preview-899--redpanda-documentation.netlify.app/docs/platform/reference/node-configuration-sample/#:~:text=level%20config%20node-,schema_registry%3A,-%23%20A%20list%20of)

Author: `/#:~:text=` is only available for Chrome unfortunately.

A solution I prefer is to change the table in https://docs.redpanda.com/docs/platform/reference/node-properties/ to be a list of properties with linkable headings. For #893 I'm updating its tunable properties table into a headings-list and can do the same for node properties.

---
title: Schema Registry
---

<head>
    <meta name="title" content="Redpanda Schema Registry | Redpanda Docs"/>
    <meta name="description" content="Overview, concepts, and tutorials of the schema registry in Redpanda."/>
</head>

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> Redpanda's schema registry provides the interface to store and manage event schemas. Producers register the schemas of data they send, and consumers retrieve the schemas of data they receive. The registry manages the different versions of schemas as they evolve, so producers and consumers can determine whether the data they exchange is incompatible.

Author: I'll add clarification. Producers and consumers register and retrieve the schemas they use from the registry. Schemas are versioned, and the registry supports configurable compatibility modes between schema versions. So when a producer or consumer makes a request to register an incompatible schema change, the registry returns an error for the incompatible schema request.

Its API endpoints support the following functionality: 

- Querying supported serialization formats.
- Registering schemas for a subject.
- Retrieving schemas of specific versions.
- Retrieving a list of subjects.
- Retrieving a list of schema versions for a subject.
- Configuring schema compatibilty checking.

## Schema

A schema is an external mechanism to describe the structure of data and its encoding.  

Using schemas provides multiple benefits:
- Enables a loosely coupled, data centric architecture that minimizes dependencies in code, between teams, and between producers and consumers.
- Provides human-readable documentation for an API.
- Verifies that data conforms to an API.
- Supports the generation of serializers for data.
- Manages the compatibility of evolving APIs, allowing new versions of services to be rolled out independently.

### Serialization format

>  A data serialization format defines how data is converted to and from the bytes that are stored, published, and received. 

Author: Rewrite: A data serialization format defines how data is converted to the bits that are stored and transmitted.

>  A data serialization format defines how data is converted to and from the bytes that are stored, published, and received. 

Reviewer 2: Should this say "bits" or "bytes" (like the original sentence)?

>  A data serialization format defines how data is converted to and from the bytes that are stored, published, and received. 

Author: I purposely changed "bytes" to "bits," because it's a more fine-grained picture of data serialization.

        ```python
        import requests
        import json
        def pretty(text):
          print(json.dumps(text, indent=2))

        base_uri = "http://localhost:8081"
        ```
3. [Install Redpanda with Docker](/docs/platform/quickstart/quick-start-docker).
4. Start Redpanda with Docker:

    ```bash
    docker network create redpanda-sr
    docker volume create redpanda-sr
    docker run \
    --pull=always \
    --name=redpanda-sr \
    --net=redpanda-sr \
    -v "redpanda-sr:/var/lib/redpanda/data" \
    -p 8081:8081 \
    -p 8082:8082 \
    -p 9092:9092 \
    --detach \
    docker.vectorized.io/vectorized/redpanda start \
    --overprovisioned \
    --smp 1 \
    --memory 1G \
    --reserve-memory 0M \
    --node-id 0 \
    --check=false \
    --pandaproxy-addr 0.0.0.0:8082 \
    --advertise-pandaproxy-addr 127.0.0.1:8082 \
    --kafka-addr 0.0.0.0:9092 \
    --advertise-kafka-addr redpanda-sr:9092
    ```

    - Once Redpanda is running, endpoints are documented at `http://localhost:8081/v1`.

### Query supported schema formats

To get the supported data serialization formats in the schema registry, make a GET request to the `/schemas/types` endpoint:

> <Tabs>

Reviewer 4: We can make sure that the user's tab choice is cached so that all tabs on the page change to the selected one.

We just need to give all tab blocks a group ID:

```
<Tabs groupId="curl-python">
```

If you POST an incompatible schema change, the request returns an error. For example, if you try to register a new schema with the `value` field's type changed from `long` to `int`, and compatibility is set to `BACKWARD`,  the request returns an error due to incompatibility:

<Tabs>
<TabItem value="curl" label="Curl" default>

```bash
curl -s \
  -X POST \
  "http://localhost:8081/subjects/sensor-value/versions" \
  -H "Content-Type: application/vnd.schemaregistry.v1+json" \
  -d '{"schema": "{\"type\":\"record\",\"name\":\"sensor_sample\",\"fields\":[{\"name\":\"timestamp\",\"type\":\"long\",\"logicalType\":\"timestamp-millis\"},{\"name\":\"identifier\",\"type\":\"string\",\"logicalType\":\"uuid\"},{\"name\":\"value\",\"type\":\"int\"}]}"}' \
  | jq
```

</TabItem>
<TabItem value="python" label="Python">

```python
sensor_schema["fields"][2]["type"] = "int"

res = requests.post(
    url=f'{base_uri}/subjects/sensor-value/versions',
    data=json.dumps({
      'schema': json.dumps(sensor_schema)
    }),
    headers={'Content-Type': 'application/vnd.schemaregistry.v1+json'}).json()
pretty(res)
```

</TabItem>
</Tabs>

The request returns this error:

```json
{
  "error_code": 409,

>   "message": "Schema being registered is incompatible with an earlier schema for subject \"{sensor-value}\""

Author: CC @BenPope

  -H "Content-Type: application/vnd.schemaregistry.v1+json" \
  -d '{"schema": "{\"type\":\"record\",\"name\":\"sensor_sample\",\"fields\":[{\"name\":\"timestamp\",\"type\":\"long\",\"logicalType\":\"timestamp-millis\"},{\"name\":\"identifier\",\"type\":\"string\",\"logicalType\":\"uuid\"},{\"name\":\"value\",\"type\":\"long\"}]}"}' \
  | jq
```

</TabItem>
<TabItem value="python" label="Python">

```python
sensor_schema = {
  "type": "record",
  "name": "sensor_sample",
  "fields": [
    {
      "name": "timestamp",
      "type": "long",
      "logicalType": "timestamp-millis"
    },
    {
      "name": "identifier",
      "type": "string",
      "logicalType": "uuid"
    },
    {
      "name": "value",
      "type": "long"
    }
  ]
}

res = requests.post(
    url=f'{base_uri}/subjects/sensor-value/versions',
    data=json.dumps({
      'schema': json.dumps(sensor_schema)
    }),
    headers={'Content-Type': 'application/vnd.schemaregistry.v1+json'}).json()
pretty(res)
```

</TabItem>
</Tabs>

If the request is successful, it returns a version `id` unique for the schema in the Redpanda cluster:

```json
{
  "id": 1
}
```

> Registering an evolved schema for an existing subject returns an incremented `id`.

Reviewer 2: How about:
```suggestion
When you register an evolved schema for an existing subject, the version `id` is incremented by 1.
```

}
```

For an example of a compatible change, register a schema with the `value` field's type changed from `long` to `double`:

<Tabs groupId="curl-python">
<TabItem value="curl" label="Curl" default>

```bash
curl -s \
  -X POST \
  "http://localhost:8081/subjects/sensor-value/versions" \
  -H "Content-Type: application/vnd.schemaregistry.v1+json" \
  -d '{"schema": "{\"type\":\"record\",\"name\":\"sensor_sample\",\"fields\":[{\"name\":\"timestamp\",\"type\":\"long\",\"logicalType\":\"timestamp-millis\"},{\"name\":\"identifier\",\"type\":\"string\",\"logicalType\":\"uuid\"},{\"name\":\"value\",\"type\":\"double\"}]}"}' \
  | jq
```

</TabItem>
<TabItem value="python" label="Python">

```python
sensor_schema["fields"][2]["type"] = "double"

res = requests.post(
    url=f'{base_uri}/subjects/sensor-value/versions',
    data=json.dumps({
      'schema': json.dumps(sensor_schema)
    }),
    headers={'Content-Type': 'application/vnd.schemaregistry.v1+json'}).json()
pretty(res)
```

</TabItem>
</Tabs>

A successful registration returns the schema's `id`:

```json
{
  "id": 2
}
```

### Delete a schema

> The schema registry API provides DELETE endpoints for deleting a schema or all schemas of a subject: 

Reviewer 2: Just a nit (see if you think it improves readability):
```suggestion
The schema registry API provides DELETE endpoints for deleting a single schema or all schemas of a subject: 
```

- `/subjects/<subject>/versions/<version>`
- `/subjects/<subject>`

A schema can be soft deleted (impermanently) or hard deleted (permanently), based on the boolean query parameter `permanent`. A soft deleted schema can be retrieved and re-registered. A hard deleted schema cannot be recovered.

#### Soft delete a schema

To soft delete a schema, make a DELETE request with the subject and version ID (where `permanent=false` is the default parameter value):

<Tabs groupId="curl-python">
<TabItem value="curl" label="Curl" default>

```bash
curl -s \
  -X DELETE \
  "http://localhost:8081/subjects/sensor-value/versions/1" \
  | jq .
```

</TabItem>
<TabItem value="python" label="Python">

```python
res = requests.delete(f'{base_uri}/subjects/sensor-value/versions/1').json()
pretty(res)
```

</TabItem>
</Tabs>

If the request is successful, it returns the ID of the soft deleted schema:

```bash
1
```

Doing a soft delete for an already deleted schema returns an error. For example:

```json
{
  "error_code": 40406,
  "message": "Subject 'sensor-value' Version 1 was soft deleted.Set permanent=true to delete permanently"
}
```

> To list subjects of soft deleted schemas, make a GET request with the `deleted` parameter set true, `/subjects?deleted=true`:

Reviewer 2: ```suggestion
To list subjects of soft-deleted schemas, make a GET request with the `deleted` parameter set true, `/subjects?deleted=true`:
```

<Tabs groupId="curl-python">
<TabItem value="curl" label="Curl" default>

```bash
curl -s \
  "http://localhost:8081/subjects?deleted=true" \
  | jq .
```

</TabItem>
<TabItem value="python" label="Python">

```python
payload = { 'deleted' : 'true' }
res = requests.get(f'{base_uri}/subjects', params=payload).json()
pretty(res)
```

</TabItem>
</Tabs>

If the request is successful, it returns all subjects, including deleted ones:

```json
[
  "sensor-value"
]
```

> To undo a soft deletion, follow the steps first to [retrieve the schema](#retrieve-a-schema-of-a-subject) then to [register the schema](#register-a-schema).

Reviewer 2: ```suggestion
To undo a soft deletion, first follow the steps to [retrieve the schema](#retrieve-a-schema-of-a-subject), then [register the schema](#register-a-schema).
```

#### Hard delete a schema

:::caution

> We don't recommend hard (permanently) deleting schemas in a production system. 

Reviewer 2: ```suggestion
Redpanda doesn't recommend hard (permanently) deleting schemas in a production system. 
```

> The DELETE APIs are provided primarily for use during the development phase, when schemas are being iterated and revised.

Reviewer 2: ```suggestion
The DELETE APIs are primarily used during the development phase, when schemas are being iterated and revised.
```

:::

> To hard delete a schema, make two DELETE requests with the second request setting the `permanent` parameter to true, `/subjects/<subject>/versions/<version>?permanent=true`:

Reviewer 2: ```suggestion
To hard delete a schema, make two DELETE requests with the second request setting the `permanent` parameter to `true`, `/subjects/<subject>/versions/<version>?permanent=true`:
```

<Tabs groupId="curl-python">
<TabItem value="curl" label="Curl" default>

```bash
curl -s \
  -X DELETE \
  "http://localhost:8081/subjects/sensor-value/versions/1" \
  | jq .
curl -s \
  -X DELETE \
  "http://localhost:8081/subjects/sensor-value/versions/1?permanent=true" \
  | jq .
```

</TabItem>
<TabItem value="python" label="Python">

```python
res = requests.delete(f'{base_uri}/subjects/sensor-value/versions/1').json()
pretty(res)
payload = { 'permanent' : 'true' }
res = requests.delete(f'{base_uri}/subjects/sensor-value/versions/1', params=payload).json()
pretty(res)
```

</TabItem>
</Tabs>

> If the requests are successful, each request return the version ID of the deleted schema:

Reviewer 2: ```suggestion
If the requests are successful, each request returns the version ID of the deleted schema:
```

```json
1
1
```

> A request for a hard deleted schema returns an error. For example:

Reviewer 2: ```suggestion
A request for a hard-deleted schema returns an error. For example:
```

```json
{
  "error_code": 40401,
  "message": "Subject 'sensor-value' not found."
}
```

### Cleanup 

> When you're finished, stop and clean up Redpanda:

Reviewer 2: Is there a better way to say this?
Maybe a brief summary of the steps (what are you stopping, and what are you removing)?

Author: Revised
